spring-batch-infrastructure/src/main/java/org/springframework/batch/item/data/builder/RepositoryItemReaderBuilder.java

```diff
@@ -0,0 +1,217 @@
+
+import java.util.List;
+import java.util.Map;
+
+import org.springframework.batch.item.ExecutionContext;
+import org.springframework.batch.item.data.RepositoryItemReader;
+import org.springframework.data.domain.Sort;
+import org.springframework.data.repository.PagingAndSortingRepository;
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+
+/**
+ * A builder implementation for the {@link RepositoryItemReader}.
+ *
+ * @author Glenn Renfro
+ * @since 4.0
+ * @see RepositoryItemReader
+ */
+
+public class RepositoryItemReaderBuilder<T> {
+
+	private PagingAndSortingRepository<?, ?> repository;
+
+	private Map<String, Sort.Direction> sorts;
+
+	private List<?> arguments;
+
+	private int pageSize = 10;
+
+	private String methodName;
+
+	private int currentItemCount;
+
+	private int maxItemCount;
+
+	private boolean saveState = true;
+
+	private String name;
+
+	/**
+	 * Arguments to be passed to the data providing method.
+	 *
+	 * @param arguments list of method arguments to be passed to the repository.
+	 * @return The current instance of the builder.
+	 * @see RepositoryItemReader#setArguments(List)
+	 */
+	public RepositoryItemReaderBuilder<T> arguments(List<?> arguments) {
+		this.arguments = arguments;
+
+		return this;
+	}
+
+	/**
+	 * Provides ordering of the results so that order is maintained between paged queries.
+	 *
+	 * @param sorts the fields to sort by and the directions.
+	 * @return The current instance of the builder.
+	 * @see RepositoryItemReader#setSort(Map)
+	 */
+	public RepositoryItemReaderBuilder<T> sorts(Map<String, Sort.Direction> sorts) {
+		this.sorts = sorts;
+
+		return this;
+	}
+
+	/**
+	 * Establish the pageSize for the generated RepositoryItemReader.
+	 *
+	 * @param pageSize The number of items to retrieve per page.
+	 * @return The current instance of the builder.
+	 * @see RepositoryItemReader#setPageSize(int)
+	 */
+	public RepositoryItemReaderBuilder<T> pageSize(int pageSize) {
+		this.pageSize = pageSize;
+
+		return this;
+	}
+
+	/**
+	 * The {@link org.springframework.data.repository.PagingAndSortingRepository}
+	 * implementation used to read input from.
+	 *
+	 * @param repository underlying repository for input to be read from.
+	 * @return The current instance of the builder.
+	 * @see RepositoryItemReader#setRepository(PagingAndSortingRepository)
+	 */
+	public RepositoryItemReaderBuilder<T> repository(PagingAndSortingRepository<?, ?> repository) {
+		this.repository = repository;
+
+		return this;
+	}
+
+	/**
+	 * Specifies what method on the repository to call. This method must take
+	 * {@link org.springframework.data.domain.Pageable} as the <em>last</em> argument.
+	 *
+	 * @param methodName name of the method to invoke.
+	 * @return The current instance of the builder.
+	 * @see RepositoryItemReader#setMethodName(String)
+	 */
+	public RepositoryItemReaderBuilder<T> methodName(String methodName) {
+		this.methodName = methodName;
+
+		return this;
+	}
```

Comment:
PR is updated allowing the user to pass in a method.  But, after updating,  I had second thoughts as to whether my solution meets the criteria of type safety.  Please take a peak though.